### ISSUE TITLE
DFT energies different compared to PySCF

### ISSUE THREAD

**Issue by manassharma07** [psi4/psi4#2911]
I am newbie to Psi4 so I am probably doing something wrong.

I am trying to calibrate Psi4 with PySCF but getting quite different DFT energies for a simple LDA calculation on Ethane.
PySCF = -78.9836765087814
Psi4    = -79.3264756599

The following is my python script
```python
import os

ncores = 8
os.environ['OMP_NUM_THREADS'] = str(ncores)
os.environ["OPENBLAS_NUM_THREADS"] = str(ncores) # export OPENBLAS_NUM_THREADS=4
os.environ["MKL_NUM_THREADS"] = str(ncores) # export MKL_NUM_THREADS=4
os.environ["VECLIB_MAXIMUM_THREADS"] = str(ncores) # export VECLIB_MAXIMUM_THREADS=4
os.environ["NUMEXPR_NUM_THREADS"] = str(ncores) # export NUMEXPR_NUM_THREADS=4
# Set the max memory for PySCF
os.environ["PYSCF_MAX_MEMORY"] = str(120000)

import psi4
psi4.set_memory(int(120000*1e6))
numpy_memory = 40
import pyscf
from pyscf import dft
# pyscf.lib.misc.num_threads(n=8)

xyzFilename = 'Ethane.xyz'

#psi4_geo = psi4.geometry(geo_txt)
with open(xyzFilename,"r") as f:
    xyz_string = f.read()

psi4_geo = psi4.core.Molecule.from_string(xyz_string, dtype='xyz',fix_symmetry='c1', fix_orientation=True,fix_com=True)

# run psi4 calculation 
psi4.core.set_num_threads(ncores)
psi4.set_options(
    {
        "scf__reference": "rks",
        "scf_type": "df",
        "scf__maxiter": 50,
        "basis": 'def2-svp',
        "df_basis_scf": 'def2-universal-jfit',
        "puream":False,
        "guess": "sad",
        "SCF_INITIAL_ACCELERATOR":"none",
        "dft_spherical_points":434,    
        "dft_radial_points" :88,        
        "diis_max_vecs" : 6
    }
)
energy = psi4.energy(name="scf", dft_functional="svwn", molecule=psi4_geo, return_wfn=False)


# run PySCF calculation
mol = pyscf.M(
    atom=xyzFilename,  # just removing the last line that was only for psi4
    basis="def2-svp",
    symmetry=False,
    verbose=4,
    cart=True,
)
# mol.max_memory = 10_000
mf = dft.rks.RKS(mol, xc="svwn").density_fit(auxbasis='def2-universal-jfit')
mf.init_guess = 'vsap'
mf.grids.level = 7
mf.conv_tol = 1e-7
mf.kernel()
print('Nuc-Nuc PySCF= ', mol.energy_nuc())
print('One electron integrals energy',mf.scf_summary['e1'])
print('Coulomb energy ',mf.scf_summary['coul'])
print('EXC ',mf.scf_summary['exc'])
```
The XYZ file for ethane is here
```
8
Ethane
C   -0.759130   0.027045  -0.000000
C    0.759130  -0.027045  -0.000000
H   -1.125990   0.306235   1.010150
H   -1.171340  -0.966695  -0.274750
H   -1.109050   0.781825  -0.735390
H    1.109040  -0.781815   0.735390
H    1.171340   0.966695   0.274760
H    1.125990  -0.306235  -1.010150
```
The following is my output:
Psi4
```
  Memory set to 111.759 GiB by Python driver.
  Threads set to 8 by Python driver.

Scratch directory: /tmp/

*** tstart() called on xxx
*** at Tue Mar 28 17:20:51 2023

   => Loading Basis Set <=

    Name: DEF2-SVP
    Role: ORBITAL
    Keyword: BASIS
    atoms 1-2 entry C          line    90 file /home/manas/psi4conda/share/psi4/basis/def2-svp.gbs
    atoms 3-8 entry H          line    15 file /home/manas/psi4conda/share/psi4/basis/def2-svp.gbs


         ---------------------------------------------------------
                                   SCF
               by Justin Turney, Rob Parrish, Andy Simmonett
                          and Daniel G. A. Smith
                              RKS Reference
                        8 Threads, 114440 MiB Core
         ---------------------------------------------------------

  ==> Geometry <==

    Molecular point group: c1
    Full point group: C1

    Geometry (in Angstrom), charge = 0, multiplicity = 1:

       Center              X                  Y                   Z               Mass
    ------------   -----------------  -----------------  -----------------  -----------------
         C           -0.759130000000     0.027045000000     0.000000000000    12.000000000000
         C            0.759130000000    -0.027045000000     0.000000000000    12.000000000000
         H           -1.125990000000     0.306235000000     1.010150000000     1.007825032230
         H           -1.171340000000    -0.966695000000    -0.274750000000     1.007825032230
         H           -1.109050000000     0.781825000000    -0.735390000000     1.007825032230
         H            1.109040000000    -0.781815000000     0.735390000000     1.007825032230
         H            1.171340000000     0.966695000000     0.274760000000     1.007825032230
         H            1.125990000000    -0.306235000000    -1.010150000000     1.007825032230

  Running in c1 symmetry.

  Rotational constants: A =      2.55497  B =      0.67557  C =      0.67557 [cm^-1]
  Rotational constants: A =  76595.94859  B =  20253.04323  C =  20253.02592 [MHz]
  Nuclear repulsion =   42.036243048604213

  Charge       = 0
  Multiplicity = 1
  Electrons    = 18
  Nalpha       = 9
  Nbeta        = 9

  ==> Algorithm <==

  SCF Algorithm Type is DF.
  DIIS enabled.
  MOM disabled.
  Fractional occupation disabled.
  Guess Type is SAD.
  Energy threshold   = 1.00e-06
  Density threshold  = 1.00e-06
  Integral threshold = 1.00e-12

  ==> Primary Basis <==

  Basis Set: DEF2-SVP
    Blend: DEF2-SVP
    Number of shells: 30
    Number of basis functions: 60
    Number of Cartesian functions: 60
    Spherical Harmonics?: false
    Max angular momentum: 2

  ==> DFT Potential <==

   => Composite Functional: SVWN <=

    Slater exchange
    Vosko, Wilk & Nusair (VWN5_RPA)
    P. A. M. Dirac, Math. Proc. Cambridge Philos. Soc. 26, 376 (1930) (10.1017/S0305004100016108)
    F. Bloch, Z. Phys. 57, 545 (1929) (10.1007/BF01340281)
    S. H. Vosko, L. Wilk, and M. Nusair, Can. J. Phys. 58, 1200 (1980) (10.1139/p80-159)

    Deriv               =              1
    GGA                 =          FALSE
    Meta                =          FALSE

    Exchange Hybrid     =          FALSE
    MP2 Hybrid          =          FALSE

   => Exchange Functionals <=

    1.0000         XC_LDA_X

   => Correlation Functionals <=

    1.0000   XC_LDA_C_VWN_RPA

   => LibXC Density Thresholds  <==

    XC_LDA_C_VWN_RPA:  1.00E-15
    XC_LDA_X:  1.00E-15

   => Molecular Quadrature <=

    Radial Scheme          =       TREUTLER
    Pruning Scheme         =           NONE
    Nuclear Scheme         =       TREUTLER

    Blocking Scheme        =         OCTREE
    BS radius alpha        =              1
    Pruning alpha          =              1
    Radial Points          =             88
    Spherical Points       =            434
    Total Points           =         292072
    Total Blocks           =           2181
    Max Points             =            256
    Max Functions          =             60
    Weights Tolerance      =       1.00E-15

   => Loading Basis Set <=

    Name: DEF2-UNIVERSAL-JFIT
    Role: JKFIT
    Keyword: DF_BASIS_SCF
    atoms 1-2 entry C          line   170 file /home/manas/psi4conda/share/psi4/basis/def2-universal-jfit.gbs
    atoms 3-8 entry H          line    17 file /home/manas/psi4conda/share/psi4/basis/def2-universal-jfit.gbs

  ==> Integral Setup <==

  DFHelper Memory: AOs need 0.007 GiB; user supplied 83.698 GiB. Using in-core AOs.

  ==> MemDFJK: Density-Fitted J/K Matrices <==

    J tasked:                   Yes
    K tasked:                    No
    wK tasked:                   No
    OpenMP threads:               8
    Memory [MiB]:             85706
    Algorithm:                 Core
    Schwarz Cutoff:           1E-12
    Mask sparsity (%):       0.0000
    Fitting Condition:        1E-10

   => Auxiliary Basis Set <=

  Basis Set: DEF2-UNIVERSAL-JFIT
    Blend: DEF2-UNIVERSAL-JFIT
    Number of shells: 60
    Number of basis functions: 194
    Number of Cartesian functions: 194
    Spherical Harmonics?: false
    Max angular momentum: 4

  Cached 100.0% of DFT collocation blocks in 0.121 [GiB].

  Minimum eigenvalue in the overlap matrix is 4.2926659058E-03.
  Reciprocal condition number of the overlap matrix is 5.8671564054E-04.
    Using symmetric orthogonalization.

  ==> Pre-Iterations <==

  SCF Guess: Superposition of Atomic Densities via on-the-fly atomic UHF (no occupation information).

   -------------------------
    Irrep   Nso     Nmo
   -------------------------
     A         60      60
   -------------------------
    Total      60      60
   -------------------------

  ==> Iterations <==

                           Total Energy        Delta E     RMS |[F,P]|

   @DF-RKS iter SAD:   -78.86898106418064   -7.88690e+01   0.00000e+00
   @DF-RKS iter   1:   -78.88636762953359   -1.73866e-02   1.49756e-02 DIIS
   @DF-RKS iter   2:   -78.27124917688414    6.15118e-01   2.18391e-02 DIIS
   @DF-RKS iter   3:   -79.32381221986704   -1.05256e+00   1.08985e-03 DIIS
   @DF-RKS iter   4:   -79.32638959495554   -2.57738e-03   2.16014e-04 DIIS
   @DF-RKS iter   5:   -79.32647363265359   -8.40377e-05   3.30059e-05 DIIS
   @DF-RKS iter   6:   -79.32647565598309   -2.02333e-06   1.23652e-06 DIIS
   @DF-RKS iter   7:   -79.32647565991985   -3.93676e-09   7.27369e-08 DIIS
  Energy and wave function converged.


  ==> Post-Iterations <==

   Electrons on quadrature grid:
      Ntotal   =   18.0000006848 ; deviation = 6.848e-07

    Orbital Energies [Eh]
    ---------------------

    Doubly Occupied:

       1A     -9.811979     2A     -9.811957     3A     -0.686974
       4A     -0.549234     5A     -0.403737     6A     -0.403737
       7A     -0.338480     8A     -0.310133     9A     -0.310133

    Virtual:

      10A      0.023031    11A      0.076058    12A      0.080886
      13A      0.080887    14A      0.110657    15A      0.110658
      16A      0.173580    17A      0.360176    18A      0.370157
      19A      0.370158    20A      0.455135    21A      0.455137
      22A      0.543892    23A      0.571369    24A      0.610757
      25A      0.610759    26A      0.613674    27A      0.623635
      28A      0.623637    29A      0.682101    30A      0.875957
      31A      1.031832    32A      1.031833    33A      1.292103
      34A      1.292104    35A      1.442476    36A      1.448632
      37A      1.662612    38A      1.675074    39A      1.675078
      40A      1.685189    41A      1.685192    42A      1.745843
      43A      1.745846    44A      1.748838    45A      1.953626
      46A      1.953628    47A      2.104265    48A      2.256310
      49A      2.256312    50A      2.286838    51A      2.457344
      52A      2.457345    53A      2.522627    54A      2.738613
      55A      2.803606    56A      2.803611    57A      2.896284
      58A      2.896288    59A      5.038202    60A      5.264241

    Final Occupation by Irrep:
              A
    DOCC [     9 ]
    NA   [     9 ]
    NB   [     9 ]

  @DF-RKS Final Energy:   -79.32647565991985

   => Energetics <=

    Nuclear Repulsion Energy =             42.0362430486042129
    One-Electron Energy =                -188.7935335064416904
    Two-Electron Energy =                  79.9597081286466675
    DFT Exchange-Correlation Energy =     -12.5288933307290371
    Empirical Dispersion Energy =           0.0000000000000000
    VV10 Nonlocal Energy =                  0.0000000000000000
    Total Energy =                        -79.3264756599198506

Computation Completed


Properties will be evaluated at   0.000000,   0.000000,   0.000000 [a0]

Properties computed using the SCF density matrix


 Multipole Moments:

 ------------------------------------------------------------------------------------
     Multipole            Electronic (a.u.)      Nuclear  (a.u.)        Total (a.u.)
 ------------------------------------------------------------------------------------

 L = 1.  Multiply by 2.5417464519 to convert [e a0] to [Debye]
 Dipole X            :          0.0000185           -0.0000189           -0.0000004
 Dipole Y            :         -0.0000204            0.0000189           -0.0000015
 Dipole Z            :         -0.0000155            0.0000189            0.0000034
 Magnitude           :                                                    0.0000037

 ------------------------------------------------------------------------------------

*** tstop() called on baldr at Tue Mar 28 17:20:52 2023
Module time:
        user time   =       6.85 seconds =       0.11 minutes
        system time =       0.17 seconds =       0.00 minutes
        total time  =          1 seconds =       0.02 minutes
Total time:
        user time   =       6.85 seconds =       0.11 minutes
        system time =       0.17 seconds =       0.00 minutes
        total time  =          1 seconds =       0.02 minutes
```

PySCF
```
[ENV] PYSCF_MAX_MEMORY 120000
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 8
[INPUT] num. electrons = 18
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry False subgroup None
[INPUT] Mole.unit = angstrom
[INPUT] Cartesian GTO integrals (6d 10f)
[INPUT] Symbol           X                Y                Z      unit          X                Y                Z       unit  Magmom
[INPUT]  1 C     -0.759130000000   0.027045000000   0.000000000000 AA   -1.434547792941   0.051107643039   0.000000000000 Bohr   0.0
[INPUT]  2 C      0.759130000000  -0.027045000000   0.000000000000 AA    1.434547792941  -0.051107643039   0.000000000000 Bohr   0.0
[INPUT]  3 H     -1.125990000000   0.306235000000   1.010150000000 AA   -2.127812718999   0.578700279756   1.908906844729 Bohr   0.0
[INPUT]  4 H     -1.171340000000  -0.966695000000  -0.274750000000 AA   -2.213511798748  -1.826788795986  -0.519202252724 Bohr   0.0
[INPUT]  5 H     -1.109050000000   0.781825000000  -0.735390000000 AA   -2.095800758449   1.477435127338  -1.389685694744 Bohr   0.0
[INPUT]  6 H      1.109040000000  -0.781815000000   0.735390000000 AA    2.095781861188  -1.477416230077   1.389685694744 Bohr   0.0
[INPUT]  7 H      1.171340000000   0.966695000000   0.274760000000 AA    2.213511798748   1.826788795986   0.519221149985 Bohr   0.0
[INPUT]  8 H      1.125990000000  -0.306235000000  -1.010150000000 AA    2.127812718999  -0.578700279756  -1.908906844729 Bohr   0.0

nuclear repulsion = 42.0362430684635
number of shells = 30
number of NR pGTOs = 92
number of NR cGTOs = 60
basis = def2-svp
ecp = {}
CPU time:         9.06


******** <class 'pyscf.df.df_jk.density_fit.<locals>.DensityFitting'> ********
method = DensityFitting-RKS
initial guess = vsap
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
SCF conv_tol = 1e-07
SCF conv_tol_grad = None
SCF max_cycles = 50
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /home/manas/temp_pyfock/PyFock/benchmarks_tests/tmpl5bjsnbu
max_memory 120000 MB (current use 564 MB)
XC library pyscf.dft.libxc version 5.2.0
    S. Lehtola, C. Steigemann, M. J. Oliveira, and M. A. Marques, SoftwareX 7, 1 (2018)
XC functionals = svwn
    P. A. M. Dirac, Math. Proc. Cambridge Philos. Soc. 26, 376 (1930)
    F. Bloch, Z. Phys. 57, 545 (1929)
    S. H. Vosko, L. Wilk, and M. Nusair, Can. J. Phys. 58, 1200 (1980)
radial grids:
    Treutler-Ahlrichs [JCP 102, 346 (1995); DOI:10.1063/1.469408] (M4) radial grids

becke partition: Becke, JCP 88, 2547 (1988); DOI:10.1063/1.454033
pruning grids: <function nwchem_prune at 0x7f820b704e50>
grids dens level: 7
symmetrized grids: False
atomic radii adjust function: <function treutler_atomic_radii_adjust at 0x7f820b704dc0>
small_rho_cutoff = 1e-07
Set gradient conv threshold to 0.000316228
tot grids = 480040
  HOMO = -0.59588105950131  LUMO = -0.145771851483052
init E= -78.2672379484169
  HOMO = 0.0562061977567527  LUMO = 0.154769986816453
cycle= 1 E= -77.2766309425052  delta_E= 0.991  |g|= 2.29  |ddm|=  4.4
  HOMO = -0.86605610499077  LUMO = -0.530054212299216
cycle= 2 E= -76.0866708577871  delta_E= 1.19  |g|= 3.32  |ddm|=  5.4
  HOMO = -0.310784522994396  LUMO = 0.033865934104578
cycle= 3 E= -78.957817824413  delta_E= -2.87  |g|= 0.349  |ddm|= 3.31
  HOMO = -0.28088985478148  LUMO = 0.0460544323558919
cycle= 4 E= -78.9829463593556  delta_E= -0.0251  |g|= 0.0521  |ddm|= 0.313
  HOMO = -0.289788725574994  LUMO = 0.040054065956731
cycle= 5 E= -78.9836641922412  delta_E= -0.000718  |g|= 0.0067  |ddm|= 0.0553
  HOMO = -0.292029983036761  LUMO = 0.038621881934318
cycle= 6 E= -78.9836755259711  delta_E= -1.13e-05  |g|= 0.00201  |ddm|= 0.00767
  HOMO = -0.292189185676856  LUMO = 0.0385039007757071
cycle= 7 E= -78.983676479896  delta_E= -9.54e-07  |g|= 0.000345  |ddm|= 0.0014
  HOMO = -0.292212979004177  LUMO = 0.0384947187953393
cycle= 8 E= -78.9836765088735  delta_E= -2.9e-08  |g|= 1.51e-05  |ddm|= 0.000283
  HOMO = -0.292216542136698  LUMO = 0.0384932286386545
Extra cycle  E= -78.9836765087814  delta_E= 9.21e-11  |g|= 2.43e-05  |ddm|= 3.39e-05
converged SCF energy = -78.9836765087814
Nuc-Nuc PySCF=  42.0362430684635
One electron integrals energy -188.76684395643963
Coulomb energy  79.92907919464169
EXC  -12.182154815446964
```

**Comment by hokru** [comment]
IIRC they will use a different VWN correlation functional as defaults. PSI4 tells you it uses `XC_LDA_C_VWN_RPA` from libxc. Can you check what pyscf uses?

**Comment by JonathonMisiewicz** [comment]
> IIRC they will use a different VWN correlation functional as defaults.

I can confirm this. To reproduce Psi's SVWN in PySCF, the exchange-correlation combination you need is `slater,vwn_rpa`. It's that `_rpa` part that PySCF doesn't do by default.

I can personally guarantee that by doing this (and setting grid pruning consistently), I've been able to get PySCF and Psi to match for 4 decimal places on TD-LDA excitation energies. It's probably even more than that, and ground state energies should have even more decimal places of agreement.

**Comment by manassharma07** [comment]
Thanks a lot for such quick and useful replies!!! @hokru @JonathonMisiewicz 

You're both right. 
I needed to use

`mf = dft.rks.RKS(mol, xc="LDA,VWNRPA").density_fit(auxbasis='def2-universal-jfit')`

or 

`mf = dft.rks.RKS(mol, xc="1, 8").density_fit(auxbasis='def2-universal-jfit')`

with PySCF. 
As @JonathonMisiewicz mentioned was indeed using `1,7`.

**Comment by JonathonMisiewicz** [comment]
Great!

If you need more precision, set the PySCF grids by a pair of numbers, rather than by grid level. That should disable PySCF's numerical integration grid pruning. Psi is not pruned, by default.

**Comment by manassharma07** [comment]
@JonathonMisiewicz Thanks a lot for the advice!

Yes for larger alkanes like Decane the energy agreement was not more 3-4 decimal places.

Based on your advice I modified the script, however, there is not much of an improvement in the agreement. In fact the total number of grid points also differ.

```python
import os

ncores = 8
os.environ['OMP_NUM_THREADS'] = str(ncores)
os.environ["OPENBLAS_NUM_THREADS"] = str(ncores) # export OPENBLAS_NUM_THREADS=4
os.environ["MKL_NUM_THREADS"] = str(ncores) # export MKL_NUM_THREADS=4
os.environ["VECLIB_MAXIMUM_THREADS"] = str(ncores) # export VECLIB_MAXIMUM_THREADS=4
os.environ["NUMEXPR_NUM_THREADS"] = str(ncores) # export NUMEXPR_NUM_THREADS=4
# Set the max memory for PySCF
os.environ["PYSCF_MAX_MEMORY"] = str(120000)

import psi4
psi4.set_memory(int(120000*1e6))
numpy_memory = 40
import pyscf
from pyscf import dft
# pyscf.lib.misc.num_threads(n=8)

xyzFilename = 'Decane_C10H22.xyz'

with open(xyzFilename,"r") as f:
    xyz_string = f.read()

psi4_geo = psi4.core.Molecule.from_string(xyz_string, dtype='xyz',fix_symmetry='c1', fix_orientation=True,fix_com=True)


# run psi4 calculation
psi4.core.set_num_threads(ncores)
psi4.set_options(
    {
        "scf__reference": "rks",
        "scf_type": "df",
        "scf__maxiter": 50,
        "basis": 'def2-svp',
        "df_basis_scf": 'def2-universal-jfit',
        "puream":False,
        "guess": "sap",
        "SCF_INITIAL_ACCELERATOR":"none",
        "dft_spherical_points":302,
        "dft_radial_points" :75,
        "diis_max_vecs" : 6,
        #"DFT_NUCLEAR_SCHEME":"becke",
        "E_CONVERGENCE":1e-7,
    }
)
energy = psi4.energy(name="scf",dft_functional="svwn", molecule=psi4_geo, return_wfn=False)
energy = 0

# run PySCF calculation
mol = pyscf.M(
    atom=xyzFilename,  # just removing the last line that was only for psi4
    basis="def2-svp",
    symmetry=False,
    verbose=4,
    cart=True,
)
# mol.max_memory = 10_000
mf = dft.rks.RKS(mol, xc="LDA,VWNRPA").density_fit(auxbasis='def2-universal-jfit')
mf.init_guess = 'vsap'
# mf.grids.level = 5
mf.verbose = 4
mf.grids.prune = None
mf.grids.atom_grid = (75, 302)
mf.grids.becke_scheme = dft.gen_grid.stratmann
mf.small_rho_cutoff = 1e-15
mf.conv_tol = 1e-7
mf.kernel()
print('Nuc-Nuc PySCF= ', mol.energy_nuc())
print('One electron integrals energy',mf.scf_summary['e1'])
print('Coulomb energy ',mf.scf_summary['coul'])
print('EXC ',mf.scf_summary['exc'])
```

Is there any further advice that you might have?

**Comment by susilehtola** [comment]
> You're both right. I needed to use
> 
> `mf = dft.rks.RKS(mol, xc="LDA,VWNRPA").density_fit(auxbasis='def2-universal-jfit')`
> 
> or
> 
> `mf = dft.rks.RKS(mol, xc="1, 8").density_fit(auxbasis='def2-universal-jfit')`

Don't use integer ids, use full libxc keywords to combine unambiguity and human readability: `xc = 'lda_x,lda_c_vwn_rpa'`. You can also use libxc identifiers in Psi4 but the syntax is a bit more verbose (you need to define a Psi4-style functional)

**Comment by susilehtola** [comment]
> Great!
> 
> If you need more precision, set the PySCF grids by a pair of numbers, rather than by grid level. That should disable PySCF's numerical integration grid pruning. Psi is not pruned, by default.

That does not suffice. In addition to setting the grid with e.g. `mf.grids.atom_grid = (100,590)` you need to turn pruning off explicitly with `mf.grids.prune = None`.

**Comment by susilehtola** [comment]
By defaults PySCF truncates basis functions to `1e-8` while the threshold is much tighter in Psi4. It can be controlled in newer PySCF as detailed in https://github.com/pyscf/pyscf/issues/1313

There are also be differences in how linear dependencies are handled in density fitting. I think Psi4 diagonalizes the Coulomb overlap `(A|B)` and throws away linearly dependent functions while PySCF directly solves the system of linear equations. In my experience, PySCF hangs much more easily in linear dependencies than Psi4.

Linear dependencies are really the weak point in PySCF overall. Orbital basis linear dependencies also are not removed by default, and one needs to do this separately, which is a common issue for practical calculations. (This should not be an issue in the def2-SVP calculation here since the basis is so small.) The linear dependency issues also pop up in other places in the code, and I had tickets open for a long time about these, but I think the last of them have now been closed so calculations with strong linear dependencies should work in the PySCF master branch, e.g. doing diatomics at negligible internuclear distances like R = 1e-4 bohr.

I would first try to reproduce the energies with exact integrals, since those numbers should match exactly. Psi4 will use density fitting to get the orbitals close to convergence for `scf_type direct`; you can do a similar thing in PySCF. For some systems PySCF will be way faster than Psi4 with exact integrals, since it handles generally contracted basis sets.

**Comment by susilehtola** [comment]
So in addition to setting `mf.small_rho_cutoff = 1e-15` in PySCF you also need to use a smaller threshold for the basis function screening.

The default density screening threshold in Psi4 comes from Libxc, which is usually `1e-15`. I think there is another keyword for controlling the basis function threshold.

I recently looked at the density threshold in finite element calculations on atoms in [arXiv:2302.06284](https://arxiv.org/abs/2302.06284). `1e-11` was sufficient to converge the total energy for all functionals to 1e-7 Hartree. The results for molecules may be different, though, since one might have larger regions of space with low electron density.

Usually one is able to get different programs to agree beyond 1 uEh in DFT total energies, when the calculations are converged with respect to all numerical parameters. I usually only look at small calculations, though, since my usual interest is just to check that a functional is implemented the same way. Here that is not an issue since libxc provides the density functional to both Psi4 and PySCF, which is a huge boon for reproducibility.

**Comment by manassharma07** [comment]
BTW the energy agreement b/w PySCF and psi4 for Ethane with the above script is up to 1e-11 a.u. so I guess, this is as good as it is gonna get.

**Comment by susilehtola** [comment]
> BTW the energy agreement b/w PySCF and psi4 for Ethane with the above script is up to 1e-11 a.u. so I guess, this is as good as it is gonna get.

So what did you change from the earlier try?

**Comment by manassharma07** [comment]
I think I didn't convey what I meant here. 

The earlier problem I reported in this issue was that the Ethane LDA energies were not matching, which got resolved by using the correct functional. But I had stated there that the energies only matched up to 6 decimal places. Later on with your and @JonathonMisiewicz 's advice I posted a new script comparing the energies for a different molecule (Decane) with the two programs. There I stated that the energy difference was 35 uH. And then as a side note, I have mentioned that for a smaller Ethane molecule, however, the difference is only 1e-11 H **_using the same script_**.
